.circleci/config.yml

```diff
@@ -1,9 +1,26 @@
-version: 2.0
-jobs:
-  build:
+version: 2.1
+executors:
+  docker-executor:
     docker:
       - image: outpostuniverse/ubuntu-18.04-gcc-sdl2-physfs-circleci
+  macos-executor:
+    macos:
+      xcode: "10.0.0"
+jobs:
+  build-linux:
+    executor: docker-executor
     steps:
-      - run: brew install physfs sdl2 sdl2_image sdl2_mixer sdl2_ttf glew
       - checkout
       - run: make -k
+  build-macos:
+    executor: macos-executor
+    steps:
+      - checkout
+      - run: brew install physfs sdl2 sdl2_image sdl2_mixer sdl2_ttf glew
```

Comment:
If the `brew install` commands were moved ahead of checkout, that would make the remaining build steps identical between platforms. That might enable use of `commands` to reduce repetition.

Reply:
- `checkout` must occur before a cache restore.
 - `brew install` and `brew link` are part of the restore/save cache block and cannot be moved outside of it.
 - See #135

Reply:
Can you provide a reference for the `checkout` restriction? That seems surprising.

Reply:
https://circleci.com/docs/2.0/configuration-reference/#section=reference
 - Shows checkout as first step in a job

https://www.youtube.com/watch?v=0-ePDGv1qLc
 - Shows checkout as first step in a job

Reply:
I didn't see any restriction mentioned in those links. Just a few examples.

In one example I saw, they were caching the source code to speed up the Git checkout. That meant putting a cache restore before the `checkout` step, and a cache save after it. In other examples they are caching dependencies installed by tooling, where the source code contains a file which lists the dependencies. In those cases the `checkout` step was (arbitrarily) placed before the cache restore. After the cache restore they then install dependencies, and save them to the cache.

In our case, we have no list of dependency packages in version control, and no tool running on MacOS to install those dependencies from a list stored in a file. Instead, we expect the dependencies to be pre-installed as part of the environment. This is much like when a developer does a one-time install of dependencies to their local machine, and then just builds repeatedly from then on. That would be matched by doing a cache restore, global environment install of dependencies, a cache save, and then finally all the needed build steps, which might include checking out or updating code, building, and testing.

----

I think we can use the [`commands`](https://circleci.com/docs/2.0/configuration-reference/#commands-requires-version-21) feature here, with two commands. A generic "build-and-test" command for all platforms, and some kind of install dependencies (with caching) command for MacOS. The `jobs` section would then just invoke one or both commands, depending on the platform.

Reply:
Found an example where caching the source is a good idea.
https://circleci.com/docs/2.0/caching/#source-caching
---
I expected there to be multiple tests performed that may be OS-dependent (Filesystem comes to mind immediately. Was also going to say DirectX but CircleCI doesn't do windows....yet) The workflow I imagined for each OS was: build on each OS, fan in to run common tests, fan out for OS-dependent tests, fan in for deploy on OS.

https://circleci.com/docs/2.0/workflows/#fan-outfan-in-workflow-example

But would you consider "any OS build breaking means the entire project is broken"? If so, then putting all the builds into one job makes sense. I don't think this is the case though.
It just means there may be one small thing wrong with that implementation for that OS and when it's fixed the one CircleCI job can be re-run.